SIP plugin: add logger callback function for processing SIP message traces in a single log callback for Sofia >= 1.13

ycherniavskyi opened the pull request:
Fixes #3365.

lminiero commented:
Thanks for reviving this! I'll review in the next few days and get back to you :+1: 
Have you tested this with both 1.12 and 1.13?

ycherniavskyi commented:
I've specifically tested on version 1.13. And in current implementation logic for version 1.12 was not changed at all.

lminiero commented on src/plugins/janus_sip.c:
```diff
@@ -1801,6 +1801,95 @@ static void janus_sip_sofia_logger(void *stream, char const *fmt, va_list ap) {
+		if(strstr(buffer, "OPTIONS") == buffer)
+			return;
+		/* Search for the Call-ID header and extract its value if it exists */
+		{
```
What's this bracket doing here? There's no condition. Please get rid of it and properly indent the following code.

ycherniavskyi replied:
I used this bracket to group the `Call-ID` finding logic and limit the scope of the `from` local variable.
I will remove it as per your request.

lminiero commented on src/plugins/janus_sip.c:
```diff
@@ -1801,6 +1801,95 @@ static void janus_sip_sofia_logger(void *stream, char const *fmt, va_list ap) {
+	char *buffer = va_arg(ap, char *);
+
+	/* Check if this is a message we need */
+	if(buffer[0] == 's' && buffer[1] == 'e' && buffer[2] == 'n' && buffer[3] == 'd' && buffer[4] == ' ') {
```
You're accessing 5 bytes of a buffer that hasn't been validated. I understand that this comes from Sofia, and so it's likely good, but still I'd rather see a `strlen` done on `buffer` before, so that we can check if there are indeed enough bytes for your check.

ycherniavskyi replied:
Understood. I will add a `strlen` check on `buffer`.

ycherniavskyi replied:
Alternatively, to maintain consistency with other parts of the code, we can use the `strstr` function instead of following the logic of the original logger handler here.

lminiero commented on src/plugins/janus_sip.c:
```diff
@@ -1801,6 +1801,95 @@ static void janus_sip_sofia_logger(void *stream, char const *fmt, va_list ap) {
+		buffer++;
+		/* If this is an OPTIONS, we don't care: drop it */
+		if(strstr(buffer, "OPTIONS") == buffer)
+			return;
```
Why? This callback is meant to pass _ALL_ SIP messages to event handlers, as long as they belong to a session. This OPTIONS may be part of a dialog.

ycherniavskyi replied:
This logic was adapted from the original `janus_sip_sofia_logger` handler function [here](https://github.com/meetecho/janus-gateway/blob/ea2799743d49938077c8267d85af1cf79777d3c5/src/plugins/janus_sip.c#L1778-L1781). I included it here to maintain consistency with the original implementation. If you approve of having different logic in logger handlers, I can remove it.

ycherniavskyi replied:
@lminiero waiting for your confirmation to remove this condition to drop OPTIONS messages.
However, it’s worth noting that in my cases, the PBX-es sends numerous automated without any meaning OPTIONS messages (roughly every 2-5 seconds) within each SIP session, likely to maintain NAT ports. This was probable the reason they were originally dropped in the logger handler.

ycherniavskyi replied:
I completely agree with you. Therefore, I will leave the current logic unchanged for now.

lminiero commented on src/plugins/janus_sip.c:
```diff
@@ -1801,6 +1801,95 @@ static void janus_sip_sofia_logger(void *stream, char const *fmt, va_list ap) {
+							count = count >= sizeof(call_id) ? sizeof(call_id) - 1 : count;
+							memcpy(call_id, start, count);
+							call_id[count] = '\0';
+
```
Unneeded empty line here.

ycherniavskyi replied:
Sure, I'll remove this empty line.
I had originally used it to visually separate these two code blocks.

lminiero reviewed (commented):
Thanks, I've added some notes inline. Revisiting the logger, I'm starting to wonder whether it's safe to use the static `call_id` as we do here: are we sure the logger is always invoked by the same thread? Have you checked this as part of your tests?

lminiero commented:
@ycherniavskyi were you waiting for more feedback from me, or have all notes been addressed in the commits that followed?

ycherniavskyi commented:
@lminiero I've addressed all your comments except for the one about dropping OPTIONS, which we decided to leave (assuming I understood you correctly). Therefore, from my perspective, this PR is ready to merge! 😊

lminiero commented:
Thanks, merging then! I'll also backport this to `0.x` :+1:
